File: .repo/system-design-mastery-module-2-database-fundamentals/0-sql-vs-nosql/1-java/src/main/java/academy/starci/sqlvsnosql/compare/CompareService.java
package academy.starci.sqlvsnosql.compare;

import academy.starci.sqlvsnosql.dto.CreateProductDto;
import academy.starci.sqlvsnosql.nosql.NosqlService;
import academy.starci.sqlvsnosql.sql.SqlService;
import org.springframework.stereotype.Service;
import java.util.*;

@Service
public class CompareService {

    private static final String[] CATEGORIES = {
        "Electronics", "Clothing", "Books", "Home & Garden", "Sports", "Toys", "Food", "Automotive"
    };

    private final SqlService sqlService;
    private final NosqlService nosqlService;

    public CompareService(SqlService sqlService, NosqlService nosqlService) {
        this.sqlService = sqlService;
        this.nosqlService = nosqlService;
    }

    public Map<String, Object> seed(int count) {
        List<CreateProductDto> dtos = new ArrayList<>(count);
        Random rand = new Random();

        for (int i = 0; i < count; i++) {
            String name = "Product-" + (i + 1);
            double price = Math.round((rand.nextDouble() * 500.0 + 1.0) * 100.0) / 100.0;
            String category = CATEGORIES[rand.nextInt(CATEGORIES.length)];
            
            Map<String, Object> metadata = new HashMap<>();
            metadata.put("sku", String.format("SKU-%06d", i + 1));
            metadata.put("weight", Math.round((rand.nextDouble() * 10.0) * 100.0) / 100.0);
            metadata.put("inStock", rand.nextDouble() > 0.2);

            dtos.add(new CreateProductDto(name, price, category, metadata));
        }

        // Clear existing
        sqlService.deleteAll();
        nosqlService.deleteAll();

        // Seed
        sqlService.createMany(dtos);
        nosqlService.createMany(dtos);

        Map<String, Object> res = new HashMap<>();
        res.put("seeded", count);
        return res;
    }

    public Map<String, Object> compare(String category) {
        // SQL Category query timing
        long start = System.nanoTime();
        var sqlCatRes = sqlService.findByCategory(category);
        double sqlCatMs = (System.nanoTime() - start) / 1_000_000.0;

        // NoSQL Category query timing
        start = System.nanoTime();
        var nosqlCatRes = nosqlService.findByCategory(category);
        double nosqlCatMs = (System.nanoTime() - start) / 1_000_000.0;

        // SQL Search query timing
        String keyword = "Product-1";
        start = System.nanoTime();
        var sqlSearchRes = sqlService.search(keyword);
        double sqlSearchMs = (System.nanoTime() - start) / 1_000_000.0;

        // NoSQL Search query timing
        start = System.nanoTime();
        var nosqlSearchRes = nosqlService.search(keyword);
        double nosqlSearchMs = (System.nanoTime() - start) / 1_000_000.0;

        Map<String, Object> response = new LinkedHashMap<>();

        Map<String, Object> findByCategory = new LinkedHashMap<>();
        findByCategory.put("sqlMs", Math.round(sqlCatMs * 100.0) / 100.0);
        findByCategory.put("nosqlMs", Math.round(nosqlCatMs * 100.0) / 100.0);
        findByCategory.put("sqlCount", sqlCatRes.size());
        findByCategory.put("nosqlCount", nosqlCatRes.size());

        Map<String, Object> search = new LinkedHashMap<>();
        search.put("sqlMs", Math.round(sqlSearchMs * 100.0) / 100.0);
        search.put("nosqlMs", Math.round(nosqlSearchMs * 100.0) / 100.0);
        search.put("sqlCount", sqlSearchRes.size());
        search.put("nosqlCount", nosqlSearchRes.size());

        response.put("findByCategory", findByCategory);
        response.put("search", search);

        return response;
    }
}

File: .repo/system-design-mastery-module-2-database-fundamentals/0-sql-vs-nosql/2-csharp/Services/CompareService.cs
using SqlVsNosql.Models;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

namespace SqlVsNosql.Services
{
    public class CompareService
    {
        private static readonly string[] Categories = {
            "Electronics", "Clothing", "Books", "Home & Garden", "Sports", "Toys", "Food", "Automotive"
        };

        private readonly SqlService _sqlService;
        private readonly NosqlService _nosqlService;

        public CompareService(SqlService sqlService, NosqlService nosqlService)
        {
            _sqlService = sqlService;
            _nosqlService = nosqlService;
        }

        public async Task<object> SeedAsync(int count)
        {
            var dtos = new List<CreateProductDto>(count);
            var rand = new Random();

            for (int i = 0; i < count; i++)
            {
                var name = $"Product-{i + 1}";
                var price = Math.Round(rand.NextDouble() * 500.0 + 1.0, 2);
                var category = Categories[rand.Next(Categories.Length)];
                
                var metadata = new JsonObject
                {
                    ["sku"] = $"SKU-{(i + 1):D6}",
                    ["weight"] = Math.Round(rand.NextDouble() * 10.0, 2),
                    ["inStock"] = rand.NextDouble() > 0.2
                };

                dtos.Add(new CreateProductDto
                {
                    Name = name,
                    Price = price,
                    Category = category,
                    Metadata = metadata
                });
            }

            await _sqlService.DeleteAllAsync();
            await _nosqlService.DeleteAllAsync();

            await _sqlService.CreateManyAsync(dtos);
            await _nosqlService.CreateManyAsync(dtos);

            return new { seeded = count };
        }

        public async Task<object> CompareAsync(string category)
        {
            var sw = new Stopwatch();

            // SQL Category
            sw.Start();
            var sqlCatRes = await _sqlService.FindByCategoryAsync(category);
            sw.Stop();
            double sqlCatMs = sw.Elapsed.TotalMilliseconds;

            // NoSQL Category
            sw.Reset();
            sw.Start();
            var nosqlCatRes = await _nosqlService.FindByCategoryAsync(category);
            sw.Stop();
            double nosqlCatMs = sw.Elapsed.TotalMilliseconds;

            // SQL Search
            var keyword = "Product-1";
            sw.Reset();
            sw.Start();
            var sqlSearchRes = await _sqlService.SearchAsync(keyword);
            sw.Stop();
            double sqlSearchMs = sw.Elapsed.TotalMilliseconds;

            // NoSQL Search
            sw.Reset();
            sw.Start();
            var nosqlSearchRes = await _nosqlService.SearchAsync(keyword);
            sw.Stop();
            double nosqlSearchMs = sw.Elapsed.TotalMilliseconds;

            return new
            {
                findByCategory = new
                {
                    sqlMs = Math.Round(sqlCatMs, 2),
                    nosqlMs = Math.Round(nosqlCatMs, 2),
                    sqlCount = sqlCatRes.Count,
                    nosqlCount = nosqlCatRes.Count
                },
                search = new
                {
                    sqlMs = Math.Round(sqlSearchMs, 2),
                    nosqlMs = Math.Round(nosqlSearchMs, 2),
                    sqlCount = sqlSearchRes.Count,
                    nosqlCount = nosqlSearchRes.Count
                }
            };
        }
    }
}
